src/esp/physics/RigidObject.cpp

```diff
@@ -213,5 +213,35 @@ Magnum::Matrix3 RigidObject::getInertiaMatrix() {
   return inertia;
 }
 
+//////////////////
+// VelocityControl
+Magnum::Matrix4 VelocityControl::integrateTransform(
+    const float dt,
+    const Magnum::Matrix4& objectTransform) {
+  // linear first
+  Magnum::Vector3 newTranslation = objectTransform.translation();
+  if (controllingLinVel) {
+    if (linVelIsLocal) {
+      newTranslation += objectTransform.rotation() *
+                        (linVel * dt);  // avoid local scaling of the velocity
+    } else {
+      newTranslation += linVel * dt;
+    }
+  }
+
+  Magnum::Matrix3 newRotationScaling = objectTransform.rotationScaling();
+  // then angular
+  if (controllingAngVel) {
+    Magnum::Vector3 globalAngVel = angVel;
+    if (angVelIsLocal) {
+      globalAngVel = objectTransform.rotation() * angVel;
+    }
+    Magnum::Quaternion q = Magnum::Quaternion::rotation(
+        Magnum::Rad{(globalAngVel * dt).length()}, globalAngVel.normalized());
+    newRotationScaling = q.toMatrix() * newRotationScaling;
```

Comment:
Just be careful.
It may cause significant numerical drift here (even with "double"). Not so sure if it would an issue in the future.

Reply:
One option to consider in the future might be switching to [TranslationRotationScalingTransformation3D](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1SceneGraph.html#a5af7f3fa56af8a3e906490e63562c8a9) that stores rotation as a quat and thus could allow for faster and more precise calculation without having to go through matrices. However extra care has to be taken as relative order of the T/R/S transformations is different than with `MatrixTransformation3D`.

Another option might be [DualQuaternionTransformation](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1SceneGraph.html#ace6339f6ce9e8983c8d02c7219419f3d) but while it reduces the amount of ALU ops even further, it is more restrictive as there's no shear or scaling.

Reply:
Storing and time stepping rotation using a unit quaternion, or exponential map is the typical way we do in rigid body simulation.

Applying DualQuaternion method to handle rigid-body dynamics is rare. It is initially introduced by Kavan to computer animation to avoid the artifacts shown in linear skinning.

Reply:
> rare

Yes, I know it got first widely used for skinning, but that's not relevant in this case. Dual quats are, at its core, an efficient way to store translation+rotation together, so why not use it like that. The only thing related to animation/skinning is the DLB operation, and that's not used here.

Reply:
Thanks for the good discussion on state representation. We should continue to keep an eye on our use cases moving forward. If we find that our current representation is lacking in speed or accuracy we can revisit the alternatives. 👍